### ISSUE TITLE
free-identifier=? raises exception when called on identifier stored in preserved syntax property

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#1495]
Apologies for the word salad title, but it pretty much summarizes the issue. I’m not totally sure if this is a bug, *per se*, or if it’s works-as-intended behavior I don’t understand, but either way, I didn’t find anything in the documentation that explains quite what I’m seeing here.

Here’s a program that reproduces the issue, made up of two files:

```racket
;;; stasher.rkt
#lang racket/base

(require (for-syntax racket/base
                     syntax/transformer)
         syntax/parse/define)

(provide foo foo-stash
         print-stash-free-id=?)

(define-syntax-parser def-id-stash
  [(_ x:id id:id)
   #:with proxy (syntax-property #'carrier 'prop (syntax-local-value #'id) #t)
   #'(define-syntax x
       (make-variable-like-transformer
        (syntax-property #'(void) 'stash
                         (syntax-property #'proxy 'prop))))])

(define-syntax-parser print-stash-free-id=?
  [(_ box-id stash)
   (define expanded-stash (local-expand #'stash 'expression '()))
   (println (free-identifier=? (syntax-local-value #'box-id)
                               (syntax-property expanded-stash 'stash)))
   #'(void)])

(define-syntax foo #'foo)
(def-id-stash foo-stash foo)

(print-stash-free-id=? foo foo-stash)
```
```racket
;;; printer.rkt
#lang racket/base

(require "stasher.rkt")

(print-stash-free-id=? foo foo-stash)
```

The first module, `stasher.rkt`, compiles fine. However, `printer.rkt` raises an error at compile-time:

```
printer.rkt:31:12: module-path-index-resolve: "self" index has no resolution
  module path index: #<module-path-index:()>
```

What’s going on here is sort of complicated, so let me break down what is happening:

  1. I define a syntax binding, `foo`, which is bound to an identifier that references itself, `#'foo`.

  2. I use a macro called `def-id-stash` to create a macro that *expands* to a piece of syntax containing a syntax property. Inside of that syntax property is the `#'foo` identifier stored in `foo`. This is done by expanding to a use of `make-variable-like-transformer`. *However*, importantly, the way the property’s value is passed is not direct—it is itself passed as a preserved syntax property.

     Specifically, `def-id-stash` creates a syntax pattern variable, `proxy`, which simply serves as a “carrier” for the identifier by attaching it as a syntax property. The expansion of `def-id-stash` uses `syntax-property` to get the value off of `#'proxy` and attach it to the piece of syntax provided to `make-variable-like-transformer`.

  3. Another macro, `print-stash-free-id=?`, simply calls `free-identifier=?` on the provided syntax binding and the “stashed” identifier produced by `def-id-stash`.

  4. When using `print-stash-free-id=?` in `stasher.rkt`, it works fine, and it prints `#t` during expansion.

  5. However, when using `print-stash-free-id=?` from another module, it produces the error mentioned earlier.

I don’t know why that error happens, and I’m not sure how to work around it, either. This case is obviously extremely convoluted, but I use something similar to it in the language I’m working on that’s based on the stuff @stchang, @AlexKnauth, and @bennn have done, so it’s in “real” code, for some definition of “real”. I’m willing to give a bit more context as to what I’m doing and why if you’d like (there is likely a much better way), but when trying to come up with a minimal reproducible test case, this is what I ended up with.

**Comment by AlexKnauth** [comment]
When I try this in Racket v6.10.1.2, it gives the `module-path-index-resolve: "self" index has no resolution` error in DrRacket, but it compiles fine (printing `#t` for stasher and `#f` for printer) from the command line.

**Comment by lexi-lambda** [comment]
You’re right. This program appears to be another bug in `module->namespace`? It works from the command line in any version of Racket newer than 6.5. Excitingly, Racket 6.5 segfaults(!) when compiling `printer.rkt`, but that’s pretty irrelevant.

My guess is my attempt to create a minimal test case for this issue reduced the program too much. Perhaps your example from the mailing list would be a better minimal test case for the bug I was *actually* trying to reproduce.

**Comment by samth** [comment]
Can you test if these programs work on racket7?

**Comment by lexi-lambda** [comment]
The program in this issue works in DrRacket in racket7 (as well as on the command-line), so I think it’s safe to say that the new expander fixes this issue as it’s currently written.

Alex’s program from [this mailing list thread](https://groups.google.com/d/msg/racket-dev/4J0uVrE3hQ4/fdFmmTWRAgAJ) still fails on racket7 with the following error:

```
expand: namespace mismatch; cannot locate module instance
  module: #<module-path-index='def[6592]>
  use phase: 0
  definition phase: 0
  for identifier: #<syntax:def.rkt:59:14 x>
  context...:
   binding->module-instance
   binding-lookup48.1
   loop_150
   /Users/alexis/gits/racket/racket7/racket/collects/syntax/parse/define.rkt:19:4: use-thing
   apply-transformer-in-context
   dispatch-transformer
   do-local-expand60.1
   /Users/alexis/gits/racket/racket7/racket/collects/syntax/wrap-modbeg.rkt:46:4: do-wrapping-module-begin
   apply-transformer-in-context
   dispatch-transformer
   loop_180
   finish_2
   [repeats 1 more time]
   pass-1-and-2-loop_0
   module-begin-k_1
   expand-module18.1
   ...
```

(I’m not terribly surprised, since Matthew’s prior comments on this issue indicate that it’s an issue with the model of syntax properties right now, not really a bug in the expander.)
